Inject Rails related configuration through Railtie

kaspth opened the pull request:
Instead of direct coupling we can inject configuration. This also spares us some setup in tests.

Work in progress. Need to consider better error messages, better docs, tests.

Would also like to tackle the `inner_logger` + `TaggedLoggerProxy.new` test setup, but unsure what to do there at the moment. Maybe we can move `new_tagged_logger` to the server?

cc @rafaelfranca @matthewd

kaspth committed:
Inject Rails' channel paths in engine.

We were explicitly referencing Rails.root in ActionCable::Server::Configuration.initialize,
thereby coupling ourselves to Rails.

Instead add `app/channels` to Rails' app paths and assign the existent files
to `channel_paths`.

Users can still append to those load paths with `<<` and `push` in `config/application.rb`.

This means we can remove the custom `Dir` lookup in `channel_paths` and the Rails
and root definitions in the tests.

kaspth committed:
Default connection class to ActionCable::Connection::Base.

Instead of depending on ApplicationCable::Connection being defined at initialize
we should inject it in the Railtie.

Thus we can kill more setup in the tests too.

matthewd commented on actioncable/lib/action_cable/engine.rb:
```diff
@@ -31,6 +31,12 @@ class Railtie < Rails::Engine # :nodoc:
 
+        channel_load_paths << Rails.root.join('app/channels')
+
+        if defined?(ApplicationCable::Connection)
```
I'm pretty sure this won't work, because of autoloading.. right?

kaspth replied:
Ah, right! Replacing with a string and `safe_constantize`.

kaspth committed:
Don't rely on the global server as a receiver.

The `WorkerTest`'s `Receiver` is imporsonating an `ActionCable::Connection::Base`, but
just delegates the logger to `ActionCable.logger`.

This creates a mismatch as the connection requires the logger to be a
`TaggedLoggerProxy`'ied logger, while the server doesn't.

Thus to ensure an exception isn't raised when the worker tries to call `tag`
other tests have to assign a proxied logger to their test server.

Instead of forcing change on other tests, have Receiver adhere to the connection
contract and use a `TaggedLoggerProxy`.

As a consequence remove more setup from the tests.

kaspth commented on actioncable/lib/action_cable/engine.rb:
```diff
@@ -31,6 +32,17 @@ class Railtie < Rails::Engine # :nodoc:
+        else
+          raise RuntimeError, <<-eow.strip_heredoc
+            Unable to find an `ApplicationCable::Connection` for Action Cable to use.
+            Check the README for instructions on where and how to set it up.
```
Quirky error message. Should be removed when we change `rails:update` to generate the `connection.rb` file.

kaspth commented on actioncable/test/client_test.rb:
```diff
@@ -12,24 +12,15 @@ class ClientTest < ActionCable::TestCase
-    server.config = ActionCable::Server::Configuration.new
-    inner_logger = Logger.new(StringIO.new).tap { |l| l.level = Logger::UNKNOWN }
-    server.config.logger = ActionCable::Connection::TaggedLoggerProxy.new(inner_logger, tags: [])
+    server.config.logger = Logger.new(StringIO.new).tap { |l| l.level = Logger::UNKNOWN }
```
Couldn't remove this without seeing circular require warnings throw in the Active Record connection management concern. Haven't yet figured out why that is.

kaspth commented on actioncable/lib/action_cable/engine.rb:
```diff
@@ -31,6 +32,10 @@ class Railtie < Rails::Engine # :nodoc:
           self.cable = Rails.application.config_for(config_path).with_indifferent_access
         end
 
+        self.connection_class = ApplicationCable::Connection
```
Killed the error message since from the point of the final release the file with this constant has been generated.

rafaelfranca replied:
Is it defined in `config`? We don't generate `app` files in that task

kaspth replied:
It's supposed to be in `app/channels/application_cable/connection.rb`.

rafaelfranca commented on actioncable/lib/action_cable/engine.rb:
```diff
@@ -31,6 +32,10 @@ class Railtie < Rails::Engine # :nodoc:
 
+        self.connection_class = ApplicationCable::Connection
+
+        channel_load_paths << Rails.root.join('app/channels')
```
We should get it from the Paths object.

rafaelfranca replied:
The main reason is that doing this engines will just work.

kaspth replied:
Ah, cool! You mean the root to join onto? I.e. `Rails.application.paths.path.join('app/channels')`?.

rafaelfranca replied:
`channel_load_paths = paths['app/channels'].existent`

rafaelfranca commented on actioncable/lib/action_cable/engine.rb:
```diff
@@ -2,6 +2,7 @@
 require "action_cable"
 require "action_cable/helpers/action_cable_helper"
 require "active_support/core_ext/hash/indifferent_access"
+require "active_support/core_ext/string/strip"
```
If you don't need the message this can be removed

kaspth replied:
Doh! Good catch.

spastorino commented:
@kaspth looks great, :+1:

rafaelfranca commented:
Fuck love this! Thank you for working on this.

kaspth commented:
@rafaelfranca awww, ❤️❤️❤️
